### PR TITLE
use etcd2 as storage-backend for federation until federation features are completely tested with etcd3

### LINKED ISSUE #39594: Federation API server is failing to process requests.
It looks like an `etcd3` compatibility issue from the initial debugging, but I haven't had a chance to do a thorough analysis.

Federation control plane creates an `etcd` container with an `etcd2` image and recent changes announced here https://groups.google.com/forum/#!topic/kubernetes-dev/WMPK5fyOorE might be the reason for the failures.

Were there any changes to generic API server to make this change happen? 

@nikhiljindal @shashidharatd, could one of you please take a look? 

cc @kubernetes/sig-federation-misc @nikhiljindal @shashidharatd @csbell @timothysc

### PR DESCRIPTION
**Pull request by shashidharatd**
**What this PR does / why we need it**: move federation etcd to etcd3

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #39594

**Special notes for your reviewer**: here is the [link](https://github.com/kubernetes/features/issues/44#issuecomment-270772674) to announcement making etcd3 as default

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by shashidharatd** [comment]
@madhusudancs @nikhiljindal please review this PR

File: federation/manifests/federation-apiserver-deployment.yaml [hunk 1]
```diff
@@ -42,9 +42,9 @@ spec:
           mountPath: /srv/kubernetes/
           readOnly: true
       - name: etcd
-        image: quay.io/coreos/etcd:v2.3.3
+        image: gcr.io/google_containers/etcd:3.0.14-alpha.1
         command:
-          - /etcd
+          - /usr/local/bin/etcd
```
Comment (madhusudancs) [inline federation/manifests/federation-apiserver-deployment.yaml]: I don't think this is sufficient. Federated Service is in beta, which means we need to provide an upgrade path. I think we should be doing what kubernetes is doing. See the manifest template here - https://github.com/kubernetes/kubernetes/blob/84df0679671e719971734be09e954c28f7febaf9/cluster/saltbase/salt/etcd/etcd.manifest#L42

File: federation/pkg/kubefed/init/init.go [hunk 2]
```diff
@@ -463,9 +463,9 @@ func createAPIServer(clientset *client.Clientset, namespace, name, image, creden
 						},
 						{
 							Name:  "etcd",
-							Image: "quay.io/coreos/etcd:v2.3.3",
+							Image: "gcr.io/google_containers/etcd:3.0.14-alpha.1",
 							Command: []string{
-								"/etcd",
+								"/usr/local/bin/etcd",
```
Comment (madhusudancs) [inline federation/pkg/kubefed/init/init.go]: Same as above.

**Review by madhusudancs — CHANGES_REQUESTED** [review]

File: federation/manifests/federation-apiserver-deployment.yaml (hunk 1 above)
Comment (shashidharatd) [inline federation/manifests/federation-apiserver-deployment.yaml]: So @madhusudancs, what is your suggestion for federation, we should make etcd3 as default, but still provide support for etcd2 or we should support etcd2 as default but optionally support etcd3?

Comment (shashidharatd) [inline federation/manifests/federation-apiserver-deployment.yaml]: @madhusudancs, I have made the federation-apiserver to use etcd2 as storage-backend.  going forward we need to migrate to etcd3 asap. also probably document the etcd migration procedure w.r.t federation.

File: federation/pkg/kubefed/init/init.go (hunk 2 above)
Comment (shashidharatd) [inline federation/pkg/kubefed/init/init.go]: updated the PR, just to solve the failing test cases for now. we need to migrate to etcd3 asap.

File: federation/manifests/federation-apiserver-deployment.yaml (hunk 1 above)
Comment (madhusudancs) [inline federation/manifests/federation-apiserver-deployment.yaml]: Fair enough, to get the tests passing. But I see that you made `etcd2` the default backend in `kubefed init`, but I don't see the corresponding change in this manifest. Our current test infra still uses this manifest. Shouldn't there be an equivalent change in this file?

Comment (shashidharatd) [inline federation/manifests/federation-apiserver-deployment.yaml]: @madhusudancs , the change is done in federation/manifests/federation-apiserver-deployment.yaml also.
The federation-apiserver now is called with `--storage-backend=etcd2` [here](https://github.com/kubernetes/kubernetes/pull/39646/files#diff-292011a5f785a3eb1dfc13fe070ab8ffR17) is the change.

**Comment by shashidharatd** [comment]
@k8s-bot unit test this

File: federation/manifests/federation-apiserver-deployment.yaml (hunk 1 above)
Comment (madhusudancs) [inline federation/manifests/federation-apiserver-deployment.yaml]: Oh weird! I am only seeing it now.

**Review by madhusudancs — APPROVED** [review]
@shashidharatd thanks! LGTM!

Also, could you please update the issue about the work that still needs to be done?

**Comment by madhusudancs** [comment]
I am adding `release-note-none` label here. I really think we should get the `etcd3` migration part done before v1.6 goes out and that should have a release note. There is no need to have one here since the one here will be superseded by the release note in the `etcd3` migration PR.

**Pull request merged**